Python/dynarray.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Dynamic array implementation.
+ */
+
+#include "pycore_dynarray.h"
+#include "pycore_pymem.h" // _PyMem_RawStrdup
+
+static inline void
+call_deallocator_maybe(_PyDynArray *array, Py_ssize_t index)
+{
+    if (array->deallocator != NULL && array->items[index] != NULL)
+    {
+        array->deallocator(array->items[index]);
+    }
+}
+
+int
+_PyDynArray_InitWithSize(_PyDynArray *array,
+                         _PyDynArray_Deallocator deallocator,
+                         Py_ssize_t initial)
+{
+    assert(array != NULL);
+    assert(initial > 0);
+    void **items = PyMem_RawCalloc(sizeof(void *), initial);
+    if (items == NULL)
+    {
+        return -1;
+    }
+
+    array->capacity = initial;
+    array->items = items;
+    array->length = 0;
+    array->deallocator = deallocator;
+
+    return 0;
+}
+
+static int
+resize_if_needed(_PyDynArray *array)
+{
+    if (array->length == array->capacity)
+    {
+        // Need to resize
+        array->capacity *= 2;
+        void **new_items = PyMem_RawRealloc(
+                                            array->items,
+                                            sizeof(void *) * array->capacity);
+        if (new_items == NULL)
+        {
+            --array->length;
+            return -1;
+        }
+
+        // XXX Zero-out the new capacity?
```

Comment:
Why is this needed?

Reply:
`realloc()` doesn't initialize memory to 0, so there's junk memory in `items`. However, it's not possible for someone to touch that junk memory unless they're manually using `items`, so it's not an issue. I can remove the comment if you want.

Reply:
I think that if someone directly touches the `items` field and something goes wrong, it's their fault. So, I'm in favor of removing this comment.